### PR TITLE
Core: persistent stores

### PR DESCRIPTION
**Pull request by Gilthoniel**
Also: change the block index logic to start at 0 instead of assuming the genesis block is the zero index.

### CODE REVIEW

File: core/store/hashtree/binprefix/mod.go [hunk 1]
```diff
@@ -60,9 +64,35 @@ func NewMerkleTree(db kv.DB, nonce Nonce) *MerkleTree {
...
+
+	return t.doUpdate(func(tx kv.WritableTx) error {
+		bucket := tx.GetBucket(t.bucket)
+
+		err := t.tree.Load(bucket)
+		if err != nil {
+			return xerrors.Errorf("failed to load: %v", err)
+		}
+
+		err = t.tree.Update(t.hashFactory, bucket)
```
Comment (nkcr) [inline core/store/hashtree/binprefix/mod.go:81]: I find  `t.tree.Load(bucket)` and `t.tree.Update(...)` to be very  meaningless. But I guess it's hard to make that code more explicit at this stage.

File: core/store/hashtree/binprefix/mod.go [hunk 2]
```diff
@@ -60,9 +64,35 @@ func NewMerkleTree(db kv.DB, nonce Nonce) *MerkleTree {
 	}
 }
 
+// Load tries to read the bucket and scan it for existing leafs and populate the
```
Comment (nkcr) [inline core/store/hashtree/binprefix/mod.go:67]: I'm glad there is the comment. I would've never understood that by reading the code!

File: core/store/hashtree/binprefix/tree.go [hunk 3]
```diff
@@ -103,6 +103,73 @@ func (t *Tree) Len() int {
...
+	})
+
+	if err != nil {
+		return xerrors.Errorf("while scanning: %v", err)
+	}
+
+	return nil
+}
+
+func (t *Tree) restore(curr TreeNode, prefix *big.Int, node *DiskNode) TreeNode {
```
Comment (nkcr) [inline core/store/hashtree/binprefix/tree.go:148]: I would add a comment about what this function does

File: core/store/hashtree/binprefix/tree.go [hunk 4]
```diff
@@ -280,6 +347,10 @@ func (n *EmptyNode) Delete(key *big.Int, b kv.Bucket) (TreeNode, error) {
 func (n *EmptyNode) Prepare(nonce []byte,
 	prefix *big.Int, b kv.Bucket, fac crypto.HashFactory) ([]byte, error) {
 
+	if len(n.hash) > 0 {
```
Comment (nkcr) [inline core/store/hashtree/binprefix/tree.go:350]: I would update the function's comment:
```
// Prepare implements mem.TreeNode. It updates the hash of the node if not 
// already set and returns the digest.
```

File: core/store/hashtree/binprefix/tree.go [hunk 5]
```diff
@@ -452,6 +525,10 @@ func (n *InteriorNode) load(node TreeNode, key *big.Int, bit uint, b kv.Bucket)
 func (n *InteriorNode) Prepare(nonce []byte,
 	prefix *big.Int, b kv.Bucket, fac crypto.HashFactory) ([]byte, error) {
 
+	if len(n.hash) > 0 {
```
Comment (nkcr) [inline core/store/hashtree/binprefix/tree.go:528]: same as previous comment, update the function's comment

File: core/store/hashtree/binprefix/tree.go [hunk 6]
```diff
@@ -614,6 +691,10 @@ func (n *LeafNode) Delete(key *big.Int, b kv.Bucket) (TreeNode, error) {
 func (n *LeafNode) Prepare(nonce []byte,
 	prefix *big.Int, b kv.Bucket, fac crypto.HashFactory) ([]byte, error) {
 
+	if len(n.hash) > 0 {
```
Comment (nkcr) [inline core/store/hashtree/binprefix/tree.go:694]: same

File: core/store/hashtree/binprefix/tree.go [hunk 7]
```diff
@@ -103,6 +103,73 @@ func (t *Tree) Len() int {
 	return counter
 }
 
+// Load scans the bucket for leafs to insert them in the tree. It will then
+// persist the tree to restore the memory depth limit.
+func (t *Tree) Load(bucket kv.Bucket) error {
```
Comment (nkcr) [inline core/store/hashtree/binprefix/tree.go:108]: less elegant, but easier to understand
```suggestion
func (t *Tree) FillFrom(bucket kv.Bucket) error {
```

File: core/store/hashtree/binprefix/mod.go (hunk 1 above)
Comment (Gilthoniel) [reply core/store/hashtree/binprefix/mod.go:81]: `Update` => `CalculateRoot`
`Load` => `LaodFromDisk`

That might help a bit.

File: core/store/hashtree/binprefix/mod.go (hunk 2 above)
Comment (Gilthoniel) [reply core/store/hashtree/binprefix/mod.go:67]: See the rename above.

File: core/store/hashtree/binprefix/tree.go (hunk 7 above)
Comment (Gilthoniel) [reply core/store/hashtree/binprefix/tree.go:108]: No that's actually a good name.

**Review by nkcr — APPROVED** [review]

**Pull request merged**